File: csharp/ToolGood.Words/internals/TrieNode3.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ToolGood.Words.internals
{
    public sealed class TrieNode3
    {
        public bool End;
        public bool HasWildcard;
        public List<int> Results;
        public Dictionary<char, TrieNode3> m_values;
        private uint minflag = uint.MaxValue;
        private uint maxflag = uint.MinValue;
        public TrieNode3 WildcardNode;


        public void Add(char c, TrieNode3 node3)
        {
            if (minflag > c) { minflag = c; }
            if (maxflag < c) { maxflag = c; }
            if (m_values == null) {
                m_values = new Dictionary<char, TrieNode3>();
            }
            m_values.Add(c, node3);
        }

        public void SetResults(int index)
        {
            if (End == false) {
                End = true;
            }
            if (Results == null) {
                Results = new List<int>();
            }
            if (Results.Contains(index) == false) {
                Results.Add(index);
            }
        }

        public bool HasKey(char c)
        {
            if (m_values==null) {
                return false;
            }
            return m_values.ContainsKey(c);
        }

        public bool TryGetValue(char c, out TrieNode3 node)
        {
            if (minflag <= (uint)c && maxflag >= (uint)c) {
                return m_values.TryGetValue(c, out node);
            }
            node = null;
            return false;
        }
    }
}

File: java/toolgood.words/src/main/java/toolgood/words/internals/TrieNode3.java
package toolgood.words.internals;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public final class TrieNode3 {
    public boolean HasWildcard;
    public List<Integer> Results;
    public HashMap<Character, TrieNode3> m_values;
    private int minflag = Integer.MAX_VALUE;
    private int maxflag = 0;
    public TrieNode3 WildcardNode;


    public TrieNode3()
    {
        // Results = new ArrayList<Integer>();
        // m_values = new HashMap<Character, TrieNode3>();
    }
    public boolean End(){
        return Results!=null;
    }

    public void Add(final char c, final TrieNode3 node3) {
        if (minflag > c) {
            minflag = c;
        }
        if (maxflag < c) {
            maxflag = c;
        }
        if(m_values==null){
            m_values = new HashMap<Character, TrieNode3>();
        }
        m_values.put(c, node3);
    }

    public void SetResults(final int index) {
        if(Results==null){
            Results = new ArrayList<Integer>();
        }
        if (Results.contains(index) == false) {
            Results.add(index);
        }
    }

    public boolean HasKey(final char c) {
        if (minflag <= c && maxflag >= c) {
            return m_values.containsKey(c);
        }
        return false;
    }

    public TrieNode3 GetValue(final char c) {
        return m_values.get(c);
    }
}